spec/requests/vms_spec.rb

```diff
@@ -82,6 +82,101 @@ def update_raw_power_state(state, *vms)
                                    [{"vendor" => "openstack"},
                                     {"vendor" => "openstack"}])
     end
+
+    def add_hardware_and_os_to_vms
+      [vm, vm1, vm2, vm_openstack, vm_openstack1, vm_openstack2].each do |vm_record|
+        cs = ComputerSystem.create
+        OperatingSystem.create(:name => "linux", :vm_or_template => vm_record, :computer_system => cs)
+        FactoryBot.create(:hardware, :vm_or_template => vm_record, :host => host)
+      end
+    end
+
+    context "with nested indirect virtual attribute ('operating_system.computer_system.created_at')" do
+      before { add_hardware_and_os_to_vms }
+      include_examples "for 'expand' query optimizations" do
+        let!(:resource)   { :vms }
+        let!(:index_url)  { api_vms_url }
+        let!(:uses)       { ["operating_systems", "computer_systems"] }
+        let!(:attributes) { "operating_system.computer_system.created_at,name" }
+      end
```

Comment:
it would be something like this.
obviously I haven't thought this through like you have and I'm completely ripping off your code
but this is may be a little easier for me to  read.
not sure about the spacing for any of this

```suggestion
      expect "efficiently queries index for subcollections/virtual_attributes"
        expect {
          get api_vms_url, :params => { :expand => "resources", :attributes => "operating_system.computer_system.created_at,name" }
        }.to make_database_queries(:count => 3, :matching => query_match( %w(vms operating_systems computer_systems)))
      end
      query_match(from_match)
        /SELECT.*FROM\s"(?:#{from_match.flatten.join("|")})"/m
      end
```

Reply:
This comment should be a part of https://github.com/ManageIQ/manageiq-api/pull/887